[Bug]: FirestoreIO writes fails: timer is set to max time (294247-01-09T04:00:54.775Z)

### What happened?

```
Caused by: java.lang.IllegalArgumentException: Cannot output with timestamp 2023-06-15T07:19:21.297Z. Output timestamps must be no earlier than the timestamp of the current input or timer (294247-01-09T04:00:54.775Z) minus the allowed skew (0 milliseconds) and no later than 294247-01-10T04:00:54.775Z. See the DoFn#getAllowedTimestampSkew() Javadoc for details on changing the allowed skew.
	org.apache.beam.runners.dataflow.worker.repackaged.org.apache.beam.runners.core.SimpleDoFnRunner.checkTimestamp(SimpleDoFnRunner.java:259)
	org.apache.beam.runners.dataflow.worker.repackaged.org.apache.beam.runners.core.SimpleDoFnRunner.access$1300(SimpleDoFnRunner.java:85)
	org.apache.beam.runners.dataflow.worker.repackaged.org.apache.beam.runners.core.SimpleDoFnRunner$DoFnProcessContext.outputWithTimestamp(SimpleDoFnRunner.java:416)
	org.apache.beam.sdk.io.gcp.firestore.FirestoreV1WriteFn$BaseBatchWriteFn$ProcessContextAdapter.output(FirestoreV1WriteFn.java:557)
	org.apache.beam.sdk.io.gcp.firestore.FirestoreV1WriteFn$BatchWriteFnWithSummary.handleWriteSummary(FirestoreV1WriteFn.java:96)
	org.apache.beam.sdk.io.gcp.firestore.FirestoreV1WriteFn$BaseBatchWriteFn.doFlush(FirestoreV1WriteFn.java:432)
	org.apache.beam.sdk.io.gcp.firestore.FirestoreV1WriteFn$BaseBatchWriteFn.flushBatch(FirestoreV1WriteFn.java:304)
	org.apache.beam.sdk.io.gcp.firestore.FirestoreV1WriteFn$BaseBatchWriteFn.processElement(FirestoreV1WriteFn.java:256)
```

### Issue Priority

Priority: 2 (default / most bugs should be filed as P2)

### Issue Components

- [ ] Component: Python SDK
- [X] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

--- a/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1WriteFn.java
+++ b/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1WriteFn.java
@@ -87,7 +87,6 @@ final class FirestoreV1WriteFn {
     @Override
     void handleWriteFailures(
         ContextAdapter<WriteSuccessSummary> context,
-        Instant timestamp,
         List<KV<WriteFailure, BoundedWindow>> writeFailures,
         Runnable logMessage) {
       throw new FailedWritesException(
@@ -97,11 +96,10 @@ final class FirestoreV1WriteFn {
     @Override
     void handleWriteSummary(
         ContextAdapter<WriteSuccessSummary> context,
-        Instant timestamp,
         KV<WriteSuccessSummary, BoundedWindow> tuple,
         Runnable logMessage) {
       logMessage.run();
-      context.output(tuple.getKey(), timestamp, tuple.getValue());
+      context.output(tuple.getKey(), tuple.getValue().maxTimestamp(), tuple.getValue());
     }
   }
 
@@ -125,19 +123,17 @@ final class FirestoreV1WriteFn {
     @Override
     void handleWriteFailures(
         ContextAdapter<WriteFailure> context,
-        Instant timestamp,
         List<KV<WriteFailure, BoundedWindow>> writeFailures,
         Runnable logMessage) {
       logMessage.run();
       for (KV<WriteFailure, BoundedWindow> kv : writeFailures) {
-        context.output(kv.getKey(), timestamp, kv.getValue());
+        context.output(kv.getKey(), kv.getValue().maxTimestamp(), kv.getValue());
       }
     }
 
     @Override
     void handleWriteSummary(
         ContextAdapter<WriteFailure> context,
-        Instant timestamp,
         KV<WriteSuccessSummary, BoundedWindow> tuple,
         Runnable logMessage) {
       logMessage.run();
@@ -274,7 +270,6 @@ final class FirestoreV1WriteFn {
                 getWriteType(write), getName(write));
         handleWriteFailures(
             contextAdapter,
-            clock.instant(),
             ImmutableList.of(
                 KV.of(
                     new WriteFailure(
@@ -466,7 +461,6 @@ final class FirestoreV1WriteFn {
         if (okCount == writesCount) {
           handleWriteSummary(
               context,
-              end,
               KV.of(new WriteSuccessSummary(okCount, okBytes), coerceNonNull(okWindow)),
               () ->
                   LOG.debug(
@@ -481,7 +475,6 @@ final class FirestoreV1WriteFn {
             int finalOkCount = okCount;
             handleWriteFailures(
                 context,
-                end,
                 ImmutableList.copyOf(nonRetryableWrites),
                 () ->
                     LOG.warn(
@@ -506,7 +499,6 @@ final class FirestoreV1WriteFn {
             if (okCount > 0) {
               handleWriteSummary(
                   context,
-                  end,
                   KV.of(new WriteSuccessSummary(okCount, okBytes), coerceNonNull(okWindow)),
                   logMessage);
             } else {
@@ -542,13 +534,11 @@ final class FirestoreV1WriteFn {
 
     abstract void handleWriteFailures(
         ContextAdapter<OutT> context,
-        Instant timestamp,
         List<KV<WriteFailure, BoundedWindow>> writeFailures,
         Runnable logMessage);
 
     abstract void handleWriteSummary(
         ContextAdapter<OutT> context,
-        Instant timestamp,
         KV<WriteSuccessSummary, BoundedWindow> tuple,
         Runnable logMessage);
 

--- a/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1WriteFn.java
+++ b/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1WriteFn.java
@@ -482,8 +482,7 @@ final class FirestoreV1WriteFn {
             int finalOkCount = okCount;
             handleWriteFailures(
                 context,
-                Preconditions.checkArgumentNotNull(okWindow).maxTimestamp()
-                ,
+                Preconditions.checkArgumentNotNull(okWindow).maxTimestamp(),
                 ImmutableList.copyOf(nonRetryableWrites),
                 () ->
                     LOG.warn(

--- a/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1WriteFn.java
+++ b/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1WriteFn.java
@@ -51,6 +51,7 @@ import org.apache.beam.sdk.transforms.DoFn;
 import org.apache.beam.sdk.transforms.display.DisplayData;
 import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
 import org.apache.beam.sdk.util.BackOffUtils;
+import org.apache.beam.sdk.util.Preconditions;
 import org.apache.beam.sdk.values.KV;
 import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.annotations.VisibleForTesting;
 import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.collect.ImmutableList;
@@ -87,6 +88,7 @@ final class FirestoreV1WriteFn {
     @Override
     void handleWriteFailures(
         ContextAdapter<WriteSuccessSummary> context,
+        Instant timestamp,
         List<KV<WriteFailure, BoundedWindow>> writeFailures,
         Runnable logMessage) {
       throw new FailedWritesException(
@@ -96,10 +98,11 @@ final class FirestoreV1WriteFn {
     @Override
     void handleWriteSummary(
         ContextAdapter<WriteSuccessSummary> context,
+        Instant timestamp,
         KV<WriteSuccessSummary, BoundedWindow> tuple,
         Runnable logMessage) {
       logMessage.run();
-      context.output(tuple.getKey(), tuple.getValue().maxTimestamp(), tuple.getValue());
+      context.output(tuple.getKey(), timestamp, tuple.getValue());
     }
   }
 
@@ -123,17 +126,19 @@ final class FirestoreV1WriteFn {
     @Override
     void handleWriteFailures(
         ContextAdapter<WriteFailure> context,
+        Instant timestamp,
         List<KV<WriteFailure, BoundedWindow>> writeFailures,
         Runnable logMessage) {
       logMessage.run();
       for (KV<WriteFailure, BoundedWindow> kv : writeFailures) {
-        context.output(kv.getKey(), kv.getValue().maxTimestamp(), kv.getValue());
+        context.output(kv.getKey(), timestamp, kv.getValue());
       }
     }
 
     @Override
     void handleWriteSummary(
         ContextAdapter<WriteFailure> context,
+        Instant timestamp,
         KV<WriteSuccessSummary, BoundedWindow> tuple,
         Runnable logMessage) {
       logMessage.run();
@@ -270,6 +275,7 @@ final class FirestoreV1WriteFn {
                 getWriteType(write), getName(write));
         handleWriteFailures(
             contextAdapter,
+            clock.instant(),
             ImmutableList.of(
                 KV.of(
                     new WriteFailure(
@@ -461,6 +467,7 @@ final class FirestoreV1WriteFn {
         if (okCount == writesCount) {
           handleWriteSummary(
               context,
+              Preconditions.checkArgumentNotNull(okWindow).maxTimestamp(),
               KV.of(new WriteSuccessSummary(okCount, okBytes), coerceNonNull(okWindow)),
               () ->
                   LOG.debug(
@@ -475,6 +482,8 @@ final class FirestoreV1WriteFn {
             int finalOkCount = okCount;
             handleWriteFailures(
                 context,
+                Preconditions.checkArgumentNotNull(okWindow).maxTimestamp()
+                ,
                 ImmutableList.copyOf(nonRetryableWrites),
                 () ->
                     LOG.warn(
@@ -499,6 +508,7 @@ final class FirestoreV1WriteFn {
             if (okCount > 0) {
               handleWriteSummary(
                   context,
+                  Preconditions.checkArgumentNotNull(okWindow).maxTimestamp(),
                   KV.of(new WriteSuccessSummary(okCount, okBytes), coerceNonNull(okWindow)),
                   logMessage);
             } else {
@@ -534,11 +544,13 @@ final class FirestoreV1WriteFn {
 
     abstract void handleWriteFailures(
         ContextAdapter<OutT> context,
+        Instant timestamp,
         List<KV<WriteFailure, BoundedWindow>> writeFailures,
         Runnable logMessage);
 
     abstract void handleWriteSummary(
         ContextAdapter<OutT> context,
+        Instant timestamp,
         KV<WriteSuccessSummary, BoundedWindow> tuple,
         Runnable logMessage);
 

--- a/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1WriteFn.java
+++ b/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1WriteFn.java
@@ -88,7 +88,6 @@ final class FirestoreV1WriteFn {
     @Override
     void handleWriteFailures(
         ContextAdapter<WriteSuccessSummary> context,
-        Instant timestamp,
         List<KV<WriteFailure, BoundedWindow>> writeFailures,
         Runnable logMessage) {
       throw new FailedWritesException(
@@ -126,12 +125,11 @@ final class FirestoreV1WriteFn {
     @Override
     void handleWriteFailures(
         ContextAdapter<WriteFailure> context,
-        Instant timestamp,
         List<KV<WriteFailure, BoundedWindow>> writeFailures,
         Runnable logMessage) {
       logMessage.run();
       for (KV<WriteFailure, BoundedWindow> kv : writeFailures) {
-        context.output(kv.getKey(), timestamp, kv.getValue());
+        context.output(kv.getKey(), kv.getValue().maxTimestamp(), kv.getValue());
       }
     }
 
@@ -275,7 +273,6 @@ final class FirestoreV1WriteFn {
                 getWriteType(write), getName(write));
         handleWriteFailures(
             contextAdapter,
-            clock.instant(),
             ImmutableList.of(
                 KV.of(
                     new WriteFailure(
@@ -482,7 +479,6 @@ final class FirestoreV1WriteFn {
             int finalOkCount = okCount;
             handleWriteFailures(
                 context,
-                Preconditions.checkArgumentNotNull(okWindow).maxTimestamp(),
                 ImmutableList.copyOf(nonRetryableWrites),
                 () ->
                     LOG.warn(
@@ -543,7 +539,6 @@ final class FirestoreV1WriteFn {
 
     abstract void handleWriteFailures(
         ContextAdapter<OutT> context,
-        Instant timestamp,
         List<KV<WriteFailure, BoundedWindow>> writeFailures,
         Runnable logMessage);
 

